torch/_functorch/vmap.py

```diff
@@ -391,19 +398,114 @@ def vmap(
+        # transpose chunk dim and flatten structure
+        # chunks_flat_args is a list of flatten args
+        chunks_flat_args = zip(*flat_args_chunks)
+        return chunks_flat_args
 
     @functools.wraps(func)
     def wrapped(*args, **kwargs):
         lazy_load_decompositions()
         _check_out_dims_is_int_or_int_pytree(out_dims, func)
         batch_size, flat_in_dims, flat_args, args_spec = _process_batched_inputs(in_dims, args, func)
+
+        if chunk_size is not None:
+            chunks_flat_args = _get_chunked_inputs(flat_args, flat_in_dims, batch_size, chunk_size)
+            return _chunked_vmap(func, flat_in_dims, chunks_flat_args,
+                                 args_spec, out_dims, randomness, **kwargs)
+
+        # If chunk_size is not specified.
         return _flat_vmap(
             func, batch_size, flat_in_dims, flat_args, args_spec, out_dims, randomness, **kwargs
         )
 
     return wrapped
 
 
+def _flatten_chunks_output(chunks_output_):
+    # chunks_output is a list of chunked outputs
+    # flatten chunked outputs:
+    flat_chunks_output = []
+    arg_spec = None
+    for output in chunks_output_:
+        flat_output, arg_specs = tree_flatten(output)
+        flat_chunks_output.append(flat_output)
+        if arg_spec is None:
+            arg_spec = arg_specs
+
+    # transpose chunk dim and flatten structure
+    # flat_output_chunks is flat list of chunks
+    flat_output_chunks = list(zip(*flat_chunks_output))
+    return flat_output_chunks, arg_spec
+
+
+def _concat_chunked_outputs(out_dims, arg_spec, flat_output_chunks):
+    # concat chunks on out_dim
+    flat_out_dims = _broadcast_to_and_flatten(out_dims, arg_spec)
+    assert len(flat_out_dims) == len(flat_output_chunks)
+    flat_output = []
+    for out_dim in flat_out_dims:
+        # Here, idx of `flat_output_chunks` is always `0`.
+        # This is because we delete the `0` index which shifts
+        # all the remaining elements to the left.
+        # Eg.
+        # >>> a = [1, 2, 3]
+        # >>> del a[0]  # a = [2, 3]
+        # >>> del a[0]  # a = [3]
+        flat_output.append(torch.cat(flat_output_chunks[0], dim=out_dim))
+        # release source data
+        del flat_output_chunks[0]
+    del flat_output_chunks
+
+    return flat_output
+
+
+# Applies vmap on chunked_input and returns concatenated output over the chunks.
+def _chunked_vmap(func, flat_in_dims, chunks_flat_args, args_spec, out_dims, randomness, **kwargs):
+    chunks_output = []
+    rs = torch.get_rng_state() if randomness == "same" else None
+    for flat_args in chunks_flat_args:
+        batch_size = _validate_and_get_batch_size(flat_in_dims, flat_args)
+        if batch_size == 0:
+            continue
+
+        if rs is not None:
+            torch.set_rng_state(rs)
+        chunks_output.append(
+            _flat_vmap(
+                func, batch_size, flat_in_dims, flat_args, args_spec, out_dims, randomness, **kwargs
+            )
+        )
```

Comment:
For jacrev(..., chunk_size=1), we should actually fallback to a for-loop instead of doing vmap with size 1. I forgot to consider this case in the jacrev PR. The reason being is that jacrev(..., chunk_size=1) should be equivalent to computing the jacobian in a for-loop, and a vmap with size 1 is not equivalent to a for loop because vmap has a bunch of limitations.

We likely want the same behavior for vmap(..., chunk_size), but I'm not completely sure about that. Thoughts?

Reply:
> we should actually fallback to a for-loop instead of doing vmap with size 1. a vmap with size 1 is not equivalent to a for loop because vmap has a bunch of limitations.

As a user, it might seem weird that `vmap(..., chunk_size=1)` works but other values don't. (Also, potentially deceptive while reading code, and if someone tries to change `chunk_size`, they will get errors). 

If we opt for this, I think it should be documented very clearly.

How about keeping it as is (using `vmap`) and see if users ask for it?

Reply:
We definitely need to change jacrev to not use vmap when chunk_size=1. Otherwise, we cannot claim parity with torch.autograd.functional.jacobian. For that API, users do set vectorize=False (the equivalent of chunk_size=1) in their code because vmap doesn't work. We can document it to make it clear that chunk_size=1 is equivalent to computing the jacobian with a for-loop instead of using vmap.

For vmap, I have less of an opinion. It is good to be consistent with whatever jacrev does, but I agree it can be confusing.

Reply:
That makes sense. Let's keep `vmap` as is. 

Will have a follow-up PR for `jacrev` soon. Is that ok?

Reply:
sounds good to me, thank you